### PR TITLE
refactor(test): second refactor of the test executor

### PR DESCRIPTION
**Pull request by jsternberg**
The previous changes to the test executor worked in one environment, but
was too difficult to refactor with the defers in the second test
environment. This changes the flow back so the entire test is executed
inside of run and handles reading the result by using a callback
function instead.

### Done checklist
- [x] docs/SPEC.md updated
- [x] Test cases written

### CODE REVIEW

File: cmd/flux/cmd/test.go [hunk 1]
```diff
@@ -801,10 +800,13 @@ func (t *TestReporter) Summarize(tests []*Test) {
...
 
-type TestSetupFunc func(ctx context.Context) (TestExecutor, error)
+type (
+	TestSetupFunc    func(ctx context.Context) (TestExecutor, error)
+	TestResultReader func(ctx context.Context, results flux.ResultIterator) error
+)
 
 type TestExecutor interface {
-	Run(pkg *ast.Package) (flux.ResultIterator, error)
+	Run(pkg *ast.Package, read TestResultReader) error
```
Comment (nathanielc) [inline cmd/flux/cmd/test.go:809]: Can you add a comment here for the purpose of the read function? Does the read function need to guarantee that the result is fully consumed? Or can it just drop it on the floor?

File: cmd/flux/cmd/test.go [hunk 2]
```diff
@@ -162,46 +162,45 @@ func (t *Test) Error() error {
...
-	if err != nil {
-		t.err = err
-		return
-	}
-	defer results.Release()
+	t.err = executor.Run(t.ast, t.execute)
 
+}
+
+func (t *Test) execute(ctx context.Context, results flux.ResultIterator) error {
```
Comment (nathanielc) [inline cmd/flux/cmd/test.go:169]: Why is this called execute when its supposed to just be reading the result? I found that a bit confusing. Maybe `consume` or just `read` is a better name?

**Review by nathanielc — CHANGES_REQUESTED** [review]
Looks good but had a few questions.

File: cmd/flux/cmd/test.go (hunk 2 above)
Comment (jsternberg) [reply cmd/flux/cmd/test.go:169]: I renamed it from `execute` and forgot to change this name.

**Review by nathanielc — APPROVED** [review]

**Pull request merged**